src/Control/Effect/Random/Internal.hs

```diff
@@ -0,0 +1,25 @@
+{-# LANGUAGE DeriveFunctor, ExistentialQuantification, StandaloneDeriving #-}
+module Control.Effect.Random.Internal
+( Random(..)
+) where
+
+import Control.Effect.Carrier
+import qualified System.Random as R (Random(..))
+
+data Random m k
+  = forall a . R.Random a => Random (a -> k)
+  | forall a . R.Random a => RandomR (a, a) (a -> k)
+  | forall a . Interleave (m a) (a -> k)
```

Comment:
Really pleased with how this turned out. Not all `RandomGen` instances have a viable `split`, but for those that do, we can trivially implement `MonadInterleave`.